text/0041-fixed-point.md

```diff
@@ -0,0 +1,104 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [amaranth-lang/rfcs#41](https://github.com/amaranth-lang/rfcs/pull/41)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Fixed point types
+
+## Summary
+[summary]: #summary
+
+Add fixed point types to Amaranth.
+
+## Motivation
+[motivation]: #motivation
+
+Fractional values in hardware are usually represented as some form of fixed point value.
+Without a first class fixed point type, the user has to manually manage how the value needs to be shifted to be represented as an integer and keep track of how that interacts with arithmetic operations.
+
+A fixed point type would encode and keep track of the precision through arithmetic operations, as well as provide standard operations for converting values to and from fixed point representations with correct rounding.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes a library addition `amaranth.lib.fixedpoint` with the following contents:
+
+`FixedPoint` is a `ShapeCastable` subclass.
+The following operations are defined on it:
+
+- `FixedPoint(f_width, /, *, signed)`: Create a `FixedPoint` with `f_width` fractional bits.
+- `FixedPoint(i_width, f_width, /, *, signed)`: Create a `FixedPoint` with `i_width` integer bits and `f_width` fractional bits.
+- `FixedPoint.cast(shape)`: Cast `shape` to a `FixedPoint` instance.
+- `.i_width`, `.f_width`, `.signed`: Width and signedness properties.
+- `.const(value)`: Create a fixed point constant from an `int` or `float`, rounded to the closest representable value.
+- `.as_shape()`: Return the underlying `Shape`.
+- `.__call__(target)`: Create a `FixedPointValue` over `target`.
+
+`Q(*args)` is an alias for `FixedPoint(*args, signed=True)`.
+
+`UQ(*args)` is an alias for `FixedPoint(*args, signed=False)`.
+
+`FixedPointValue` is a `ValueCastable` subclass.
+The following operations are defined on it:
+
+- `FixedPointValue(shape, target)`: Create a `FixedPointValue` with `shape` over `target`.
+- `FixedPointValue.cast(value)`: Cast `value` to a `FixedPointValue`.
+- `.i_width`, `.f_width`, `.signed`: Width and signedness properties.
+- `.shape()`: Return the `FixedPoint` this was created from.
+- `.as_value()`: Return the underlying value.
+- `.eq(value)`: Assign `value`.
+  - If `value` is a `FixedPointValue`, the precision will be extended or rounded as required.
+  - If `value` is an `int` or `float`, the value will be rounded to the closest representable value.
```

Comment:
What happens if you assign 1024 to a Q8.8 value?

Reply:
`Signal(Q(8, 8)).eq(1024)` would effectively be `Signal(signed(17)).eq(1024 << 8)`. (Or `signed(16)` depending on Q notation.)

Speaking of overflow, I figure fixed point overflow should behave like regular integer overflow. Saturating math feels orthogonal to this RFC and could be added through a later RFC that handles both integer and fixed point values.